File: client/connection.go
// Connection

package prc_client

import (
	"fmt"
	"strconv"
	"strings"
	"sync"
	"time"

	simple_rpc_message "github.com/AgustinSRG/go-simple-rpc-message"
	"github.com/gorilla/websocket"
)

// Period to send HEARTBEAT messages to the client
const HEARTBEAT_MSG_PERIOD_SECONDS = 30

// Pending request
type PendingRequest struct {
	// Request type
	requestType string

	// Parallel request limit
	limit uint32
}

// Connection to a PRC server
type Connection struct {
	// Client
	cli *Client

	// Configuration
	config *ClientConfig

	// Mutex for the struct
	mu *sync.Mutex

	// True if the client is connected
	connected bool

	// Socket
	socket *websocket.Conn

	// Wait group to prevent multiple connections
	closeWaitGroup *sync.WaitGroup

	// Pending request to send on connection
	pendingRequests map[uint64]*PendingRequest

	// Pending request counts
	pendingRequestCounts map[string]int
}

func NewConnection(cli *Client, config *ClientConfig) *Connection {
	return &Connection{
		cli:                  cli,
		config:               config,
		mu:                   &sync.Mutex{},
		connected:            false,
		socket:               nil,
		closeWaitGroup:       nil,
		pendingRequests:      make(map[uint64]*PendingRequest),
		pendingRequestCounts: make(map[string]int),
	}
}

// Opens the connection
func (conn *Connection) Connect() {
	conn.mu.Lock()

	if conn.connected {
		return
	}

	conn.connected = true

	conn.mu.Unlock()

	go conn.runConnectionLoop()
}

// Closes the connection
func (conn *Connection) Close() {
	conn.mu.Lock()

	// Clear

	conn.pendingRequests = make(map[uint64]*PendingRequest)

	conn.mu.Unlock()

	// Close

	wg := conn.closeInternal()

	if wg != nil {
		wg.Wait()
	}
}

// Closes (internal)
func (conn *Connection) closeInternal() *sync.WaitGroup {
	conn.mu.Lock()
	defer conn.mu.Unlock()

	if !conn.connected {
		return conn.closeWaitGroup
	}

	conn.connected = false
	conn.closeWaitGroup = &sync.WaitGroup{}
	conn.closeWaitGroup.Add(1)

	if conn.socket != nil {
		conn.socket.Close()
	}

	return conn.closeWaitGroup
}

// Checks if the client is closed
func (conn *Connection) IsClosed() bool {
	conn.mu.Lock()
	defer conn.mu.Unlock()

	return !conn.connected
}

// Call after the connection closes
func (conn *Connection) afterClose() {
	conn.mu.Lock()
	defer conn.mu.Unlock()

	if conn.closeWaitGroup != nil {
		conn.closeWaitGroup.Done()
		conn.closeWaitGroup = nil
	}

	conn.connected = false
	conn.socket = nil
}

// Call when connected
// Returns false if the connection must be closed
func (conn *Connection) onConnected(socket *websocket.Conn) bool {
	conn.mu.Lock()
	defer conn.mu.Unlock()

	conn.socket = socket

	// Send pending requests

	for id, req := range conn.pendingRequests {
		msg := simple_rpc_message.RPCMessage{
			Method: "START-REQUEST",
			Params: map[string]string{
				"Request-ID":    fmt.Sprint(id),
				"Request-Type":  req.requestType,
				"Request-Limit": fmt.Sprint(req.limit),
			},
			Body: "",
		}

		conn.socket.WriteMessage(websocket.TextMessage, []byte(msg.Serialize()))
	}

	// Send pending request counts

	for rType := range conn.pendingRequestCounts {
		msg := simple_rpc_message.RPCMessage{
			Method: "GET-REQUEST-COUNT",
			Params: map[string]string{
				"Request-Type": rType,
			},
			Body: "",
		}

		conn.socket.WriteMessage(websocket.TextMessage, []byte(msg.Serialize()))
	}

	return conn.connected
}

// Runs connection loop
func (conn *Connection) runConnectionLoop() {
	defer conn.afterClose()

	for {
		closed := conn.IsClosed()

		if closed {
			return
		}

		url, err := conn.config.GetFullConnectionUrl()

		if err != nil {
			return
		}

		socket, _, err := websocket.DefaultDialer.Dial(url, nil)

		if err != nil {
			if conn.config.ErrorHandler != nil {
				conn.config.ErrorHandler.OnConnectionError(err)
			}

			if conn.config.RetryConnectionDelay == 0 {
				time.Sleep(DEFAULT_RETRY_CONNECTION_DELAY)
			} else if conn.config.RetryConnectionDelay > 0 {
				time.Sleep(conn.config.RetryConnectionDelay)
			}

			continue
		}

		// Set connection

		isConnected := conn.onConnected(socket)

		if !isConnected {
			socket.Close()
			return
		}

		go conn.sendHeartbeatMessages(socket)

		// Read messages and close

		conn.readIncomingMessages(socket)
	}
}

// Reads incoming messages
func (conn *Connection) readIncomingMessages(socket *websocket.Conn) {
	defer socket.Close()

	for {
		err := socket.SetReadDeadline(time.Now().Add(HEARTBEAT_MSG_PERIOD_SECONDS * 2 * time.Second))

		if err != nil {
			if !conn.IsClosed() && conn.config.ErrorHandler != nil {
				conn.config.ErrorHandler.OnConnectionError(err)
			}
			return
		}

		mt, message, err := socket.ReadMessage()

		if err != nil {
			if !conn.IsClosed() && conn.config.ErrorHandler != nil {
				conn.config.ErrorHandler.OnConnectionError(err)
			}
			return
		}

		if mt != websocket.TextMessage {
			continue
		}

		parsedMessage := simple_rpc_message.ParseRPCMessage(string(message))

		switch strings.ToUpper(parsedMessage.Method) {
		case "ERROR":
			if conn.config.ErrorHandler != nil {
				conn.config.ErrorHandler.OnServerError(parsedMessage.GetParam("Error-Code"), parsedMessage.GetParam("Error-Message"))
			}
		case "START-REQUEST-ACK":
			conn.ReceiveStartRequestAck(&parsedMessage)
		case "REQUEST-COUNT":
			conn.ReceiveRequestCount(&parsedMessage)
		}
	}
}

// Periodically sends heartbeat messages
func (conn *Connection) sendHeartbeatMessages(socket *websocket.Conn) {
	for {
		time.Sleep(HEARTBEAT_MSG_PERIOD_SECONDS * time.Second)

		// Send heartbeat message
		msg := simple_rpc_message.RPCMessage{
			Method: "HEARTBEAT",
			Params: nil,
			Body:   "",
		}

		err := socket.WriteMessage(websocket.TextMessage, []byte(msg.Serialize()))

		if err != nil {
			return
		}
	}
}

// Send a message to the module
func (conn *Connection) Send(msg *simple_rpc_message.RPCMessage) {
	conn.mu.Lock()
	defer conn.mu.Unlock()

	if conn.socket != nil {
		conn.socket.WriteMessage(websocket.TextMessage, []byte(msg.Serialize()))
	}
}

// Sends START-REQUEST message
func (conn *Connection) sendStartRequest(id uint64, rType string, limit uint32) {
	msg := simple_rpc_message.RPCMessage{
		Method: "START-REQUEST",
		Params: map[string]string{
			"Request-ID":    fmt.Sprint(id),
			"Request-Type":  rType,
			"Request-Limit": fmt.Sprint(limit),
		},
		Body: "",
	}

	conn.Send(&msg)
}

// Sends END-REQUEST message
func (conn *Connection) sendEndRequest(id uint64) {
	msg := simple_rpc_message.RPCMessage{
		Method: "END-REQUEST",
		Params: map[string]string{
			"Request-ID": fmt.Sprint(id),
		},
		Body: "",
	}

	conn.Send(&msg)
}

// Sends GET-REQUEST-COUNT message
func (conn *Connection) sendGetRequestCount(rType string) {
	msg := simple_rpc_message.RPCMessage{
		Method: "GET-REQUEST-COUNT",
		Params: map[string]string{
			"Request-Type": rType,
		},
		Body: "",
	}

	conn.Send(&msg)
}

// Starts request, either by sending a START-REQUEST message or waiting for connection
func (conn *Connection) StartRequest(id uint64, rType string, limit uint32) {
	conn.mu.Lock()

	conn.pendingRequests[id] = &PendingRequest{
		requestType: rType,
		limit:       limit,
	}

	conn.mu.Unlock()

	conn.sendStartRequest(id, rType, limit)
}

// Ends a request, by sending the END-REQUEST message
func (conn *Connection) EndRequest(id uint64) {
	conn.mu.Lock()

	delete(conn.pendingRequests, id)

	conn.mu.Unlock()

	conn.sendEndRequest(id)
}

// Receives message: START-REQUEST-ACK
func (conn *Connection) ReceiveStartRequestAck(msg *simple_rpc_message.RPCMessage) {
	idStr := msg.GetParam("Request-ID")

	id, err := strconv.ParseUint(idStr, 10, 64)

	if err != nil {
		if conn.config.ErrorHandler != nil {
			conn.config.ErrorHandler.OnServerError("PROTOCOL_ERROR", "Server send an invalid Request-Id parameter for message START-REQUEST-ACK")
		}
		return
	}

	limitedStr := strings.ToUpper(msg.GetParam("Request-Limit-Reached"))

	limited := limitedStr == "TRUE"

	conn.cli.receiveRequestAck(id, limited)
}

// Sends GET-REQUEST-COUNT message to get the request count for a specific type
func (conn *Connection) GetRequestCount(rType string) {
	conn.mu.Lock()

	c := conn.pendingRequestCounts[rType]

	if c == 0 {
		conn.pendingRequestCounts[rType] = 1
	} else {
		conn.pendingRequestCounts[rType] = c + 1
	}

	conn.mu.Unlock()

	conn.sendGetRequestCount(rType)
}

// Call after a request count is done, either by receiving the response or due to timeout
func (conn *Connection) RequestCountDone(rType string) {
	conn.mu.Lock()

	c := conn.pendingRequestCounts[rType]

	if c > 0 {
		if c == 1 {
			delete(conn.pendingRequestCounts, rType)
		} else {
			conn.pendingRequestCounts[rType] = c - 1
		}

	}

	conn.mu.Unlock()
}

// Receives REQUEST-COUNT message
func (conn *Connection) ReceiveRequestCount(msg *simple_rpc_message.RPCMessage) {
	reqType := msg.GetParam("Request-Type")
	reqCountStr := msg.GetParam("Request-Count")

	reqCount, err := strconv.ParseUint(reqCountStr, 10, 32)

	if err != nil {
		if conn.config.ErrorHandler != nil {
			conn.config.ErrorHandler.OnServerError("PROTOCOL_ERROR", "Server send an invalid Request-Count parameter for message REQUEST-COUNT")
		}
		return
	}

	conn.cli.receiveRequestCount(reqType, uint32(reqCount))
}

File: client-js/src/connection.ts
// Connection

"use strict";

import { PRCClient } from "./client";
import { DEFAULT_RETRY_CONNECTION_DELAY, getFullConnectionUrl, PRCClientConfig } from "./config";
import { RawData, WebSocket } from "ws";
import { PendingRequest } from "./utils";
import { WebsocketMessage, makeMessage, parseMessage, getMessageParam } from "./message";

const HEARTBEAT_MSG_PERIOD_SECONDS = 30;

/**
 * Connection to the PRC server
 */
export class PRCConnection {
    // Connection index
    public index: number;

    // Configuration
    private config: PRCClientConfig;

    // Client reference
    private client: PRCClient;

    // True if closed
    private closed: boolean;

    // Socket
    private socket: WebSocket | null;

    // True if connection is open
    private connected: boolean;

    // Map of pending requests
    private pendingRequests: Map<number, PendingRequest>;

    // Map of pending request counts
    private pendingRequestCounts: Map<string, number>;

    // Interval to send heartbeat messages
    private heartBeatInterval?: NodeJS.Timeout;

    // Timeout to retry the connection
    private retryConnectionTimeout?: NodeJS.Timeout;

    // Timestamp of the last received heartbeat
    private lastReceivedHeartbeat: number;

    /**
     * Connection constructor
     * @param client Client
     * @param config Configuration
     * @param index Connection index
     */
    constructor(client: PRCClient, config: PRCClientConfig, index: number) {
        this.index = index;
        this.client = client;
        this.config = config;
        this.closed = true;
        this.connected = false;
        this.socket = null;
        this.pendingRequests = new Map();
        this.pendingRequestCounts = new Map();
        this.lastReceivedHeartbeat = Date.now();
    }

    /**
     * Starts the connection
     */
    public connect() {
        if (!this.closed) {
            return; // Already connected
        }

        this.closed = false;
        this.createConnection();
    }

    /**
     * Closes the connection
     */
    public close() {
        this.closed = true;

        if (this.heartBeatInterval) {
            clearInterval(this.heartBeatInterval);
            this.heartBeatInterval = undefined;
        }

        if (this.retryConnectionTimeout) {
            clearTimeout(this.retryConnectionTimeout);
            this.retryConnectionTimeout = undefined;
        }

        if (this.socket) {
            this.socket.close();
        }
    }

    /**
     * Creates the websocket connection
     */
    private createConnection() {
        let url: string;

        try {
            url = getFullConnectionUrl(this.config);
        } catch (ex) {
            this.closed = true;
            if (this.config.onConnectionError) {
                this.config.onConnectionError(ex);
            }
            return;
        }

        this.socket = new WebSocket(url);

        this.socket.on("open", this.onOpen.bind(this));
        this.socket.on("close", this.onClose.bind(this));
        this.socket.on("message", this.onMessage.bind(this));

        this.socket.on("error", err => {
            if (this.config.onConnectionError) {
                this.config.onConnectionError(err);
            }
        });
    }

    /**
     * Sends a message
     * @param msg The message to send
     */
    public send(msg: WebsocketMessage) {
        if (this.socket && this.connected) {
            this.socket.send(makeMessage(msg));
        }
    }

    /**
     * Called when the connection is established
     */
    private onOpen() {
        this.connected = true;

        if (this.heartBeatInterval) {
            clearInterval(this.heartBeatInterval);
            this.heartBeatInterval = undefined;
        }

        this.heartBeatInterval = setInterval(this.sendHeartBeat.bind(this), HEARTBEAT_MSG_PERIOD_SECONDS * 1000);

        // Send pending requests

        this.pendingRequests.forEach((pr, rid) => {
            this.send({
                type: "START-REQUEST",
                args: {
                    "Request-ID": rid + "",
                    "Request-Type": pr.requestType,
                    "Request-Limit": pr.limit + "",
                },
            });
        });

        this.pendingRequestCounts.forEach((_count, rType) => {
            this.send({
                type: "GET-REQUEST-COUNT",
                args: {
                    "Request-Type": rType,
                },
            });
        });
    }

    /**
     * Called when the connection is closed
     * @param code Code
     * @param reason Close reason
     */
    private onClose(code: number, reason: Buffer) {
        this.socket = null;
        this.connected = false;

        if (this.heartBeatInterval) {
            clearInterval(this.heartBeatInterval);
            this.heartBeatInterval = undefined;
        }

        if (this.retryConnectionTimeout) {
            clearTimeout(this.retryConnectionTimeout);
            this.retryConnectionTimeout = undefined;
        }

        if (!this.closed) {
            // Retry the connection
            this.retryConnectionTimeout = setTimeout(this.createConnection.bind(this), this.config.retryConnectionDelay || DEFAULT_RETRY_CONNECTION_DELAY);

            // Inform of the connection error
            if (this.config.onConnectionError) {
                this.config.onConnectionError(new Error("Lost connection. Code: " + code + ", Reason: " + reason));
            }
        }
    }

    /**
     * Sends HEARTBEAT message
     * Also checks for the server last HEARTBEAT message
     */
    private sendHeartBeat() {
        this.send({
            type: "HEARTBEAT",
        });

        if (Date.now() - this.lastReceivedHeartbeat >= (HEARTBEAT_MSG_PERIOD_SECONDS * 2 * 1000)) {
            if (this.socket) {
                this.socket.close();
            }
        }
    }

    /**
     * Called on received message
     * @param data Message data
     * @param isBinary True if binary
     */
    private onMessage(data: RawData, isBinary: boolean) {
        if (isBinary) {
            return;
        }

        const msg = parseMessage(data.toString("utf-8"));

        switch (msg.type) {
            case "ERROR":
                if (this.config.onServerError) {
                    this.config.onServerError(getMessageParam(msg, "Error-Code"), getMessageParam(msg, "Error-Message"));
                }
                break;
            case "HEARTBEAT":
                this.lastReceivedHeartbeat = Date.now();
                break;
            case "START-REQUEST-ACK":
                this.receiveStartRequestAck(msg);
                break;
            case "REQUEST-COUNT":
                this.receiveRequestCount(msg);
                break;
        }
    }

    /**
     * Sends the PRC server a message to indicate the start of a request
     * @param id Request ID
     * @param rType The request Type
     * @param limit Max number of allowed parallel requests for the specified request type
     */
    public startRequest(id: number, rType: string, limit: number) {
        this.pendingRequests.set(id, {
            requestType: rType,
            limit: limit,
        });

        this.send({
            type: "START-REQUEST",
            args: {
                "Request-ID": id + "",
                "Request-Type": rType,
                "Request-Limit": limit + "",
            },
        });
    }

    /**
     * Indicates the PRC server the ending of a request
     * @param id Request ID
     */
    public endRequest(id: number) {
        this.pendingRequests.delete(id);

        this.send({
            type: "END-REQUEST",
            args: {
                "Request-ID": id + "",
            },
        });
    }

    /**
     * Receives message: START-REQUEST-ACK
     * @param msg The message
     */
    private receiveStartRequestAck(msg: WebsocketMessage) {
        const id = parseInt(getMessageParam(msg, "Request-ID"), 10);

        if (isNaN(id)) {
            if (this.config.onServerError) {
                this.config.onServerError("PROTOCOL_ERROR", "Server send an invalid Request-Id parameter for message START-REQUEST-ACK");
            }
            return;
        }

        const limited = getMessageParam(msg, "Request-Limit-Reached").toUpperCase() === "TRUE";

        this.client.receiveRequestAck(id, limited);
    }

    /**
     * Sends a GET-REQUEST-COUNT message to the PRC server to count parallel requests
     * @param rType The request type
     */
    public getRequestCount(rType: string) {
        if (this.pendingRequestCounts.has(rType)) {
            const c = this.pendingRequestCounts.get(rType);
            this.pendingRequestCounts.set(rType, c + 1);
        } else {
            this.pendingRequestCounts.set(rType, 1);
        }

        this.send({
            type: "GET-REQUEST-COUNT",
            args: {
                "Request-Type": rType,
            },
        });
    }

    /**
     * Call after a request count is done
     * @param rType The request type
     */
    public requestCountDone(rType: string) {
        if (this.pendingRequestCounts.has(rType)) {
            const c = this.pendingRequestCounts.get(rType);

            if (c === 1) {
                this.pendingRequestCounts.delete(rType);
            } else {
                this.pendingRequestCounts.set(rType, c - 1);
            }
        }
    }

    /**
     * Receives message: REQUEST-COUNT
     * @param msg The message
     */
    private receiveRequestCount(msg: WebsocketMessage) {
        const rType = getMessageParam(msg, "Request-Type");
        const count = parseInt(getMessageParam(msg, "Request-Type"), 10);

        if (isNaN(count)) {
            if (this.config.onServerError) {
                this.config.onServerError("PROTOCOL_ERROR", "Server send an invalid Request-Count parameter for message REQUEST-COUNT");
            }
            return;
        }

        this.client.receiveRequestCount(rType, count);
    }
}

